src/pelicanfs/core.py

```diff
@@ -0,0 +1,258 @@
+            updatedUrl = metalist[0][0]
+            metalist = metalist[1:]
+            # Timeout response in seconds - the default response is 5 minutes
+            timeout = aiohttp.ClientTimeout(total=2)
+            async with aiohttp.ClientSession() as session:
+                async with session.get(updatedUrl, timeout=timeout) as resp:
+                    try:
+                        resp.status
+                    except (aiohttp.client_exceptions.ClientConnectorError, FileNotFoundError, asyncio.TimeoutError):
+                        continue
+                    break
+        if len(metalist) == 0:
+            # No working cache was found
+            raise RuntimeError
+        
+        return updatedUrl
+
+    
+    def _dirlist_dec(func):
+        """
+        Decorator function which, when given a namespace location, get the url for the dirlist location from the headers
+        and uses that url for the given function.
+
+        This is for functions which need to list information in the origin directories such as "find", "isdir", "ls"
+        """
+        async def wrapper(self, *args, **kwargs):
+            path = args[0]
+            parsedUrl = urllib.parse.urlparse(path)
+            headers = await self.get_director_headers(parsedUrl.path)
+            dirlistloc = dir_header_parser.get_dirlist_loc(headers)
+            if dirlistloc == None:
+                raise RuntimeError
+            listUrl = dirlistloc + "/" + parsedUrl.path
+            result = await func(self, listUrl, *args[1:], **kwargs)
+            return result
+        return wrapper
+
+    @_dirlist_dec
+    async def _ls(self, path, detail=True, **kwargs):
+        return await self.httpFileSystem._ls(path, detail, **kwargs)
+
+    @_dirlist_dec
+    async def _isdir(self, path):
+        return await self.httpFileSystem._isdir(path)
+    
+    @_dirlist_dec
+    async def _find(self, path, maxdepth=None, withdirs=False, **kwargs):
+        return await self.httpFileSystem._find(path, maxdepth, withdirs, **kwargs)
+    
+    # Not using a decorator because it requires a yield
+    async def _walk(self, path, maxdepth=None, on_error="omit", **kwargs):
+        parsedUrl = urllib.parse.urlparse(path)
+        headers = await self.get_director_headers(parsedUrl.path)
+        dirlistloc = dir_header_parser.get_dirlist_loc(headers)
+        if dirlistloc == "":
+            raise RuntimeError
+        listUrl = dirlistloc + "/" + path
+        async for _ in self.httpFileSystem._walk(listUrl, maxdepth, on_error, **kwargs):
+                yield _
+
+
+    def _open(
+        self,
+        path,
+        mode="rb",
+        block_size=None,
+        autocommit=None,  # XXX: This differs from the base class.
+        cache_type=None,
+        cache_options=None,
+        size=None,
+        **kwargs,
+    ):    
+        loop = asyncio.get_event_loop()
+        cache_url = loop.run_until_complete(self.get_working_cache(path))
```

Comment:
Does this cause the function to not be asynchronous?  Shouldn't this cause blocking?

Reply:
The _open in httpfsspec isn't asynchronous to begin with, which is why we need to call get_working_cache this way as it's calling an async function from a non-async function. There is a separate open_async that I still need to implement that will be.